core/src/main/java/org/opensearch/sql/datasource/DataSourceLoaderCacheImpl.java

```diff
@@ -0,0 +1,57 @@
+package org.opensearch.sql.datasource;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.opensearch.sql.datasource.model.DataSource;
+import org.opensearch.sql.datasource.model.DataSourceMetadata;
+import org.opensearch.sql.datasource.model.DataSourceType;
+import org.opensearch.sql.storage.DataSourceFactory;
+
+/**
+ * Default implementation of DataSourceLoaderCache. This implementation
+ * utilizes Google Guava Cache {@link Cache} for caching DataSource objects
+ * against {@link DataSourceMetadata}. The getOrLoadDataSource also contains
+ * the logic of invalidating the cache objects with same datasourcename.
+ */
+public class DataSourceLoaderCacheImpl implements DataSourceLoaderCache {
+  private final Map<DataSourceType, DataSourceFactory> dataSourceFactoryMap;
+  private final Cache<DataSourceMetadata, DataSource> dataSourceCache;
+
+  /**
+   * DataSourceLoaderCacheImpl constructor.
+   *
+   * @param dataSourceFactorySet set of {@link DataSourceFactory}.
+   */
+  public DataSourceLoaderCacheImpl(Set<DataSourceFactory> dataSourceFactorySet) {
+    this.dataSourceFactoryMap = dataSourceFactorySet.stream()
+        .collect(Collectors.toMap(DataSourceFactory::getDataSourceType, f -> f));
+    ;
+    this.dataSourceCache = CacheBuilder.newBuilder().build();
+  }
+
+  @Override
+  public DataSource getOrLoadDataSource(DataSourceMetadata dataSourceMetadata) {
+    DataSource dataSource = this.dataSourceCache.getIfPresent(dataSourceMetadata);
+    if (dataSource == null) {
+      invalidateCacheWithDataSourceName(dataSourceMetadata.getName());
```

Comment:
scenario
1) Created datasource prometheusA. Cache will store datasource object against metadata.
2) Now customer updates some properties of the metadata.
3) Now if we fetch Datasource from cache with updated metadata, it will be null. But there is an old Datasource object lying against old metadata object, we are invalidating entries with the same name.